gqlalchemy/instance_runner.py

```diff
@@ -0,0 +1,153 @@
+        self,
+        host: str = "127.0.0.1",
+        port: int = 7687,
+        username: str = "",
+        password: str = "",
+        encrypted: bool = False,
+    ) -> None:
+        self.host = host
+        self.port = port
+        self.username = username
+        self.password = password
+        self.encrypted = encrypted
+        self.config = {}
+        self.proc_mg = None
+
+    def set_config(self, config: Dict[str, Union[str, int, bool]]):
+        self.config.update(config)
+
+    def connect(self):
+        self.memgraph = Memgraph(self.host, self.port, self.username, self.password, self.encrypted)
+        if not self.is_running():
+            raise ConnectionError("The Memgraph process probably died.")
+        return self.memgraph
+
+    @abstractmethod
+    def is_running(self):
+        pass
+
+    @abstractmethod
+    def stop(self):
+        pass
+
+    @abstractmethod
+    def start(self, restart: bool = False, binary_path: Optional[str] = ""):
+        pass
+
+
+class MemgraphInstanceBinary(MemgraphInstance):
+    def __init__(self, binary_path="/usr/lib/memgraph/memgraph", user: str = "memgraph", **data):
+        super().__init__(**data)
+        self.binary_path = binary_path
+        self.user = user
+
+    def start(self, restart=False):
+        if not restart and self.is_running():
+            return
+        self.stop()
+
+        args_mg = f"{self.binary_path }" + (" ").join([f"{k} {v}" for k, v in self.config])
+        args_mg += f" --bolt-address {self.host} --bolt-port {self.port}"
+        args_mg = (f"sudo -u {self.user} " if self.user != "" else "") + args_mg
+        print(args_mg)
+        self.proc_mg = subprocess.Popen(args_mg, shell=True, preexec_fn=os.setsid)
+        wait_for_port(self.host, self.port)
+
+        self.connect()
+        return self.memgraph
+
+    def stop(self):
+        if not self.is_running():
+            return
+        os.killpg(os.getpgid(self.proc_mg.pid), signal.SIGTERM)
+        self.proc_mg.wait()
+
+    def is_running(self):
+        if self.proc_mg is None:
+            return False
+        if self.proc_mg.poll() is not None:
+            return False
+        return True
+
+
+class MemgraphInstanceDocker(MemgraphInstance):
+    def __init__(self, docker_image: DockerImage = DockerImage.MEMGRAPH, docker_tag: str = "latest", **data):
+        super().__init__(**data)
+        self.docker_image = docker_image
+        self.docker_tag = docker_tag
+        self.client = docker.from_env()
+        self.container = None
+
+    def start(self, restart=False):
+        if not restart and self.is_running():
+            return
+        self.stop()
+
+        self.container = self.client.containers.run(
+            self.docker_image.value + ":" + self.docker_tag + (" ").join([f"{k} {v}" for k, v in self.config]),
+            detach=True,
+            ports={f"7687/tcp": self.port},
+        )
+        wait_for_port(self.host, self.port)
+
+        self.connect()
```

Comment:
I don't really see a better module for this function than the `instance_runner`.  
I could add it to the `utilities` module, but as far as I know, it's better to keep it here as the functionality is intended for checking running Memgraph instances.  Any suggestions? @BorisTasevski

Reply:
What Buda proposes is that `connect()` call is removed from `start()` method, because it is counter-intuitive and does not comply with single responsibility principle. If the method is named `start`, it should only start, not connect also. Add different(i.e. `start_and_connect())` method for this